### PR TITLE
util: implement collation interface for `utf8_general_ci` and `utf8mb4_general_ci`

### LINKED ISSUE #14630: charset: implement collation interface for `utf8_general_ci` and `utf8mb4_general_ci` in TiDB
## Feature Request

**Describe the feature you'd like:**
We decide to implement the first 2 case-insensitive collations in TiDB, `utf8_general_ci` and `utf8mb4_general_ci` are chosen because they're the default collations of character set `utf8` and `utf8mb4` before MySQL 8.0, they're widely used by MySQL users.

**Describe alternatives you've considered:**
In MySQL 8.0, collation `utf8mb4_0900_ai_ci` is added and replaced `utf8mb4_general_ci` as the default collation of `utf8mb4`. Since it is new, the usage of it is not as wide as `utf8mb4_general_ci`, so we decide to implement it in the future.

**Teachability, Documentation, Adoption, Migration Strategy:**
Please see TiDB RFC #14574 for more details on collations support in TiDB.

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 
Close #14630

### What is changed and how it works?
Implement collation interface for `utf8_general_ci` and `utf8mb4_general_ci`
Add a function GetCollatorByID.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

**Comment by breeswish** [comment]
@andylokandy Remember to update the TiKV implementation

File: util/collate/collate.go [hunk 1]
```diff
@@ -54,6 +64,14 @@ func (bc *binCollator) Key(str string) []byte {
 
 func init() {
 	collatorMap = make(map[string]Collator)
+	collatorIDMap = make(map[int]Collator)
 
 	collatorMap["binary"] = &binCollator{}
+	collatorMap["utf8mb4_general_ci"] = &generalCICollator{}
+	collatorMap["utf8_general_ci"] = &generalCICollator{}
+
+	// See parser/charset/charset.go for more information about the IDs
```
Comment (bb7133) [inline util/collate/collate.go]:
```suggestion
	// See https://github.com/pingcap/parser/blob/master/charset/charset.go for more information about the IDs
```

File: util/collate/general_ci.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+		return -1
+	} else if i > 0 {
+		return 1
+	}
+	return 0
+}
+
+// Compare implement Collator interface.
+func (gc *generalCICollator) Compare(a, b string) int {
+	aRune := []rune(a)
```
Comment (bb7133) [inline util/collate/general_ci.go]: My concern is converting & compare the rune one by one would be faster than this, can you make a benchmark to make a comparison?

File: util/collate/collate.go [hunk 3]
```diff
@@ -39,6 +40,15 @@ func GetCollator(collate string) Collator {
 	return ctor
 }
 
+// GetCollatorByID get the collator according to id, it will return the binary collator if the corresponding collator doesn't exist.
+func GetCollatorByID(id int) Collator {
```
Comment (winoros) [inline util/collate/collate.go]: I found that, there's some collation information in parser's `mysql/charset.go`. Should we uniform them?

Comment (winoros) [inline util/collate/collate.go]: I found that, there's some collation information in parser's `mysql/charset.go`. Should we uniform them?

**Comment by wjhuang2016** [comment]
Here is Benchmark for 
```
func (gc *generalCICollator) Compare(a, b string) int {
	aRune := []rune(a)
	bRune := []rune(b)
	aLen := len(aRune)
	bLen := len(bRune)

	for i := 0; i < aLen && i < bLen; i++ {
		cmp := int(convertRune(aRune[i])) - int(convertRune(bRune[i]))
		if cmp != 0 {
			return sign(cmp)
		}
	}

	return sign(aLen - bLen)
}
```

100Diff means comparing two different strings with length 100
100eEqual means comparing two eqal strings with length 100
Simple5 means comparing two strings with length 5
50Equal50Diff means comparing two string with length 100 and their first 50 chars are equal.
```
BenchmarkComapre100Diff-4               1000000000               0.0556 ns/op
BenchmarkComapre100equal-4              1000000000               0.0947 ns/op
BenchmarkComapreSimple5-4               1000000000               0.00757 ns/op
BenchmarkComapre50Equal50Diff-4         1000000000               0.0711 ns/op

```

And here is the Benchmark for
```
func (gc *generalCICollator) Compare(a, b string) int {
	for len(a) > 0 && len(b) > 0 {
		r1, size1 := utf8.DecodeRuneInString(a)
		r2, size2 := utf8.DecodeRuneInString(b)

		cmp := int(convertRune(r1)) - int(convertRune(r2))
		if cmp != 0 {
			return sign(cmp)
		}
		a = a[size1:]
		b = b[size2:]
	}
	return sign(len(a) - len(b))
}
```
```
BenchmarkComapre100Diff-4               1000000000               0.00407 ns/op
BenchmarkComapre100equal-4              1000000000               0.178 ns/op
BenchmarkComapreSimple5-4               1000000000               0.00425 ns/op
BenchmarkComapre50Equal50Diff-4         1000000000               0.0921 ns/op
```

File: util/collate/collate.go [hunk 4]
```diff
@@ -18,16 +18,17 @@ import (
...
+	collatorMap   map[string]Collator
+	collatorIDMap map[int]Collator
 )
 
 // Collator provides functionality for comparing strings for a given
 // collation order.
 type Collator interface {
 	// Compare returns an integer comparing the two strings. The result will be 0 if a == b, -1 if a < b, and +1 if a > b.
-	Compare(a, b string) int
+	Compare(a, b string, padLen int) int
```
Comment (qw4990) [inline util/collate/collate.go]: There may be other arguments about padding like `padChar`, `ignorePad` and etc, so how about using a struct here which is easier to add some new arguments in the future.

**Pull request merged**